### PR TITLE
Emit app port and added get*Version() functions.

### PR DESCRIPTION
**Pull request by ddd-mtl**
Reads the `APP_WS_PORT:` output from holochain runner and emit value back to client.
The code makes sure IS_READY and app port has been emitted before resolving promise.

### CODE REVIEW

**Comment by Connoropolous** [comment]
nice to have your contributions here @ddd-mtl , these are things I've been meaning to tighten up!

wouldn't it have to have waited for app ws to be ready anyways? it wasn't working like that?

**Comment by Connoropolous** [comment]
I will note that I'm aware that handling when holochain crashes for various reasons is hard and hasn't been yet handled nicely in this library

**Comment by ddd-mtl** [comment]
Once an app is installed, the runner doesn't care about the appWs option you give it. So I need to retrieve whatever appWs its currently using and tell my web ui to use that.
Similar thing with the appId. The runner just gives you the first app installed it finds, whatever option you feed it.

**Comment by Connoropolous** [comment]
"Once an app is installed, the runner doesn't care about the appWs option you give it. So I need to retrieve whatever appWs its currently using and tell my web ui to use that." this is true, though if you told it to use that port, isn't it safe to assume that you know what the port is?

File: src/options.ts [hunk 1]
```diff
@@ -27,8 +27,8 @@ export function constructOptions(options: HolochainRunnerOptions): string[] {
   if (options.membraneProof) {
     optionsArr = optionsArr.concat(['--membrane-proof', options.membraneProof])
   }
-  if (options.boostrapUrl) {
-    optionsArr = optionsArr.concat(['--bootstrap-url', options.boostrapUrl])
+  if (options.bootstrapUrl) {
```
Comment (Connoropolous) [inline src/options.ts]: oh, shoot. my bad!

**Comment by ddd-mtl** [comment]
> "Once an app is installed, the runner doesn't care about the appWs option you give it. So I need to retrieve whatever appWs its currently using and tell my web ui to use that." this is true, though if you told it to use that port, isn't it safe to assume that you know what the port is?

Probably but you need to remember what port you told it to use the first time.
Also the wanted port could already be taken. Haven't tried it but at best it gives back a different port, worst case it crashs.

**Comment by Connoropolous** [comment]
Yah... I can attest that in the case where the port is already taken holochain crashes

**Comment by Connoropolous** [comment]
I dont think there can be any harm in listening for the port it is using, but I know that its a guarantee that it'll be the port you asked for. 

I suppose the exception is passing in a 0. I do not know if this is possible. In that case it could (but does it?) give you back an open port that it attached to.

**Comment by ddd-mtl** [comment]
Don't know with the runner, but thats what I do with Snapmail. I give holochain a 0 for app port and write in the conductor-config the port actually used (for future use). 
Useful for running multiple instance of your happ.

**Comment by Connoropolous** [comment]
well, we do just call `add_app_interface` for the conductor, so it's just a passthrough, however, I am noting about that code that it would not display the value picked by the conductor as the APP_WS_PORT, but the value passed in, such as `0`
https://github.com/Sprillow/holochain-runner/blob/eb9631e5568c3668d7681fe2b10f570464db3a31/src/embedded_runner.rs#L193-L204
we could easily adjust that. I am assuming that `add_app_interface` does return the selected port

**Comment by Connoropolous** [comment]
Actually... I am wondering about this: 
"I give holochain a 0 for app port and write in the conductor-config the port actually used (for future use)."
I just looked in holochain source code and can't see any evidence of app port accepting a `0` value. 

admin port does. Is that what you were referring to in your above quote?

**Comment by ddd-mtl** [comment]
I just tried it out:

If I give `appWsPort: 0` it will connect to `8888`
If I give `appWsPort: 1` it will connect to `1` even though it doesn't work.

**Comment by Connoropolous** [comment]
Can you tell me a bit about what you did with error handling in the latest commit? I would like to understand it quite well, and document it

**Review by Connoropolous — APPROVED** [review]

**Comment by Connoropolous** [comment]
published to npm as v0.0.25

**Pull request merged**